lib/services/local/generic/generic153_test.go

```diff
@@ -0,0 +1,277 @@
+		return nil, trace.BadParameter(err.Error())
+	}
+
+	if r.Metadata == nil {
+		r.Metadata = &headerv1.Metadata{}
+	}
+
+	if cfg.ID != 0 {
+		//nolint:staticcheck // SA1019. Deprecated, but still needed.
+		r.Metadata.Id = cfg.ID
+	}
+	if cfg.Revision != "" {
+		r.Metadata.Revision = cfg.Revision
+	}
+	if !cfg.Expires.IsZero() {
+		r.Metadata.Expires = timestamppb.New(cfg.Expires)
+	}
+	return &r, nil
+}
+
+// TestGenericCRUD153 tests backend operations with the generic service.
+func TestGenericCRUD153(t *testing.T) {
+	ctx := context.Background()
+
+	memBackend, err := memory.New(memory.Config{
+		Context: ctx,
+		Clock:   clockwork.NewFakeClock(),
+	})
+	require.NoError(t, err)
+
+	const backendPrefix = "generic_prefix"
+
+	service, err := NewService153[*testResource153](memBackend,
+		"generic resource",
+		backendPrefix,
+		marshalResource153,
+		unmarshalResource153)
+	require.NoError(t, err)
+
+	// Create a couple test resources.
+	r1 := newTestResource153("r1")
+	r2 := newTestResource153("r2")
+
+	// sanity check: marshal/unmarshal round trips
+	r1data, err := marshalResource153(r1)
+	require.NoError(t, err)
+	r1clone, err := unmarshalResource153(r1data)
+	require.NoError(t, err)
+	require.Equal(t, r1, r1clone)
+
+	// Initially we expect no resources.
+	out, nextToken, err := service.ListResources(ctx, 200, "")
+	require.NoError(t, err)
+	require.Empty(t, nextToken)
+	require.Empty(t, out)
+
+	// Create both resources.
+	r1, err = service.CreateResource(ctx, r1)
+	require.NoError(t, err)
+	r2, err = service.CreateResource(ctx, r2)
+	require.NoError(t, err)
+
+	require.NotEmpty(t, r1.GetMetadata().GetRevision())
+	require.NotEmpty(t, r2.GetMetadata().GetRevision())
+	require.NotEqual(t, r1.GetMetadata().GetRevision(), r2.GetMetadata().GetRevision())
+
+	// Fetch all resources using paging default.
+	out, nextToken, err = service.ListResources(ctx, 0, "")
+	require.NoError(t, err)
+	require.Empty(t, nextToken)
+	require.NotEmpty(t, out)
+	require.Equal(t, []*testResource153{r1, r2}, out)
+
+	// Fetch a paginated list of resources.
+	paginatedOut := make([]*testResource153, 0, 2)
+	numPages := 0
+	for {
+		numPages++
+		out, nextToken, err = service.ListResources(ctx, 1, nextToken)
+		require.NoError(t, err)
+
+		paginatedOut = append(paginatedOut, out...)
+		if nextToken == "" {
+			break
+		}
+	}
+
+	require.Equal(t, 2, numPages)
+	require.Equal(t, []*testResource153{r1, r2}, paginatedOut)
+
+	// Fetch a list of all resources
+	allResources, err := service.GetResources(ctx)
+	require.NoError(t, err)
+	require.Equal(t, paginatedOut, allResources)
+
+	// Fetch a specific service provider.
+	r, err := service.GetResource(ctx, r2.GetMetadata().GetName())
+	require.NoError(t, err)
+	require.Equal(t, r2, r)
+
+	// Try to fetch a resource that doesn't exist.
+	_, err = service.GetResource(ctx, "doesnotexist")
+	require.ErrorIs(t, err, trace.NotFound(`generic resource "doesnotexist" doesn't exist`))
```

Comment:
ErrorIs checks with complex errors are often brittle, I've had problems with various checks like this lately. Use more resilient assertions, for example `require.True(t, trace.IsNotFound(err))` is much less likely to break.

Same for others.

Reply:
We care about both type and error message, which is why `require.ErrorsIs` was used here. Also, this is parallels the other test which we have for legacy resources.

Reply:
Yeah, we've gone through this one many times now. I'll resolve.